ledger/participant-integration-api/src/main/scala/platform/store/dao/events/ACSReader.scala

```diff
@@ -107,32 +119,97 @@ class FilterTableACSReader(
       .map(toIdSource)
       .pipe(mergeSort[Long])
       .statefulMapConcat(statefulDeduplicate)
-      .grouped(pageSize)
-      .map(_.toVector)
+      .via(
+        BatchN(
+          maxBatchSize = pageSize,
+          maxBatchCount = acsFetchingparallelism + 1,
+        )
+      )
       .async
+      .addAttributes(
+        Attributes.inputBuffer(initial = acsFetchingparallelism, max = acsFetchingparallelism)
+      )
       .mapAsync(acsFetchingparallelism)(fetchAcs)
   }
 }
 
 private[events] object FilterTableACSReader {
+  private val logger = ContextualizedLogger.get(this.getClass)
+
   case class Filter(party: Party, templateId: Option[Identifier])
 
+  case class IdQuery(fromExclusiveEventSeqId: Long, pageSize: Int)
+
+  case class IdQueryConfiguration(
+      minPageSize: Int,
+      maxPageSize: Int,
+  )
+
+  object IdQueryConfiguration {
+    def apply(
+        maxIdPageSize: Int,
+        idPageWorkingMemoryBytes: Int,
+        filterSize: Int,
+        idPageBufferSize: Int,
+    )(implicit loggingContext: LoggingContext): IdQueryConfiguration = {
+      val LowestIdPageSize = 10
+      // Approximation how many index entries can be present in one btree index leaf page (fetching smaller than this only adds round-trip overhead, without boiling down to smaller disk read)
+      // Experiments show party_id, template_id index has 244 tuples per page, wildcard party_id index has 254 per page (with default fill ratio for BTREE Index)
+      // Picking a lower number is for accommodating pruning, deletions, index bloat effect, which boil down to lower tuple per page ratio.
+      val RecommendedMinIdPageSize = 200
+      val calculatedMaxIdPageSize =
+        idPageWorkingMemoryBytes
+          ./(8) // IDs stored in 8 bytes
+          ./(
+            idPageBufferSize + 1
+          ) // for each filter we need one page fetched for merge sorting, and additional pages might reside in the buffer
+          ./(filterSize)
```

Comment:
Would be interesting to see what are the number of filters (filterSize) that results in `calculatedMaxIdPageSize` dip below `LowestIdPageSize` and `RecommendedMinIdPageSize`

Reply:
Scratch that. I found the number if the included tests.

Reply:
LowestIdPageSize=10 -> buffer=1, filteSize=655360
RecommendedIdPageSize=200 -> buffer=1, filteSize=32768